File: docs/languages/_data/examples/java/Showcase.java
import java.util.ArrayList;
import java.util.List;
import java.util.Scanner;

/**
 * Пример-витрина: Java 21.
 *
 * Одна программа, в которой прокомментированы конструкции, соответствующие
 * концепциям каталога. Маркеры {@code --8<--} размечают секции для сайта и не
 * влияют на компиляцию. Собрать: {@code javac Showcase.java && java Showcase}.
 */
public class Showcase {

    // --8<-- [start:req-8-2]
    // 8.2. Глобальных переменных нет: любая переменная — член класса или
    // локальная. Эквивалент глобального состояния — статическое поле.
    static int counter = 0;

    static void bump() {
        counter++;
    }
    // --8<-- [end:req-8-2]

    // --8<-- [start:variants-9]
    // 9. Подпрограммы объявляются только как методы класса; вложенных методов
    // нет. Обход — лямбда или локальный класс внутри метода.
    static int outer(int x) {
        java.util.function.IntUnaryOperator inner = y -> x + y; // лямбда видит x
        return inner.applyAsInt(1);
    }
    // --8<-- [end:variants-9]

    // --8<-- [start:variants-7]
    // 7. Перегрузка методов есть: выбор по статическим типам аргументов.
    static String describe(int value) {
        return "целое " + value;
    }

    static String describe(double value) {
        return "дробное " + value;
    }
    // --8<-- [end:variants-7]

    // --8<-- [start:variants-8]
    // 8. Передача параметров — всегда по значению (JLS 8.4.1), но значение
    // ссылочного типа — это ссылка на объект: изменение объекта видно
    // вызывающему, перепривязка параметра — нет.
    static void appendAndRebind(List<Integer> items, int n) {
        items.add(n);              // видно снаружи: тот же объект
        n = n + 1;                 // не видно: копия примитива
        items = new ArrayList<>(); // не видно: локальная ссылка
    }
    // --8<-- [end:variants-8]

    // --8<-- [start:req-4]
    // 4. Ввод-вывод — классы стандартной библиотеки: System.out для вывода,
    // Scanner (или BufferedReader) для ввода. Встроенных функций нет.
    static int readInt(Scanner in) {
        return in.nextInt();
    }
    // --8<-- [end:req-4]

    public static void main(String[] args) {
        // --8<-- [start:variants-1]
        // 1. Объявление явное — тип перед именем. С версии 10 для локальных
        // переменных допустим `var`: тип выводится, но объявление остаётся.
        int total = 0;
        var name = "ЯПИС"; // String
        // --8<-- [end:variants-1]

        // --8<-- [start:variants-2]
        // 2. Преобразование типов явное — оператор приведения `(int) x`;
        // неявное — только расширяющее (int → long → double) и boxing.
        int count = Integer.parseInt("42");
        double ratio = count / 5.0;   // int → double неявно (расширение)
        int truncated = (int) ratio;  // double → int только явно
        // --8<-- [end:variants-2]

        // --8<-- [start:variants-3]
        // 3. Присваивание одиночное; `a = b = 0` — цепочка одиночных,
        // множественного `a, b = c, d` нет.
        int a, b;
        a = b = 1;
        int tmp = a;
        a = b + 1;
        b = tmp;
        // --8<-- [end:variants-3]

        // --8<-- [start:variants-4]
        // 4. Область видимости ограничивают и методы, и блоки `{ }`.
        if (count > 0) {
            String insideBlock = "видна только здесь";
            System.out.println(insideBlock);
        }
        // System.out.println(insideBlock); // ошибка компиляции
        // --8<-- [end:variants-4]

        // --8<-- [start:variants-5]
        // 5. Маркер блока явный — фигурные скобки; для одного оператора
        // их можно опустить, но это отдельный оператор, а не блок.
        for (int i = 0; i < 3; i++) {
            total += i;
        }
        // --8<-- [end:variants-5]

        // --8<-- [start:variants-6]
        // 6. if / else if / else и многовариантный switch. С версии 14
        // switch — также выражение со стрелочными ветками без fall-through;
        // с версии 21 — сопоставление с образцом по типам.
        String size;
        if (total > 100) {
            size = "много";
        } else if (total > 10) {
            size = "средне";
        } else {
            size = "мало";
        }
        String label = switch (a) {
            case 1 -> "единица";
            case 2 -> "двойка";
            default -> "что-то другое";
        };
        Object obj = ratio;
        String kind = switch (obj) {
            case Integer i -> "Integer " + i;
            case Double d -> "Double " + d;
            default -> "другой тип";
        };
        System.out.println(size + ", " + label + ", " + kind);
        // --8<-- [end:variants-6]

        // --8<-- [start:req-7-3]
        // 7.3. Цикл for в двух формах: счётный (в стиле C) и по коллекции.
        for (int i = 0; i < 10; i += 2) {
            System.out.print(i + " ");
        }
        for (char ch : name.toCharArray()) {
            System.out.print(ch + " ");
        }
        System.out.println();
        // --8<-- [end:req-7-3]

        // --8<-- [start:req-7-2-until]
        // 7.2. Цикла until нет; эквивалент — while с отрицанием условия.
        int n = 0;
        while (!(n >= 3)) {
            n++;
        }
        // --8<-- [end:req-7-2-until]

        // --8<-- [start:req-7-2-do-while]
        // Цикл do-while есть: тело выполняется хотя бы раз.
        do {
            n--;
        } while (n > 0);
        // --8<-- [end:req-7-2-do-while]

        System.out.println(describe(count) + ", " + describe(ratio) + ", " + outer(41) + ", " + truncated);
        List<Integer> items = new ArrayList<>();
        appendAndRebind(items, 5);
        System.out.println(items); // [5]
        bump();
        System.out.println(counter + " " + b);

        if (args.length > 0) {
            try (Scanner in = new Scanner(System.in)) {
                System.out.println(readInt(in));
            }
        }
    }
}

File: docs/languages/_data/examples/python/showcase.py
"""Пример-витрина: Python 3.

Одна программа, в которой прокомментированы конструкции, соответствующие
независимым концепциям каталога. Старые маркеры `--8<--` приняты через
legacy_slugs; они размечают секции для сайта и не влияют на выполнение.
"""

from __future__ import annotations

import sys


# --8<-- [start:req-8-2]
# 8.2. Глобальная область видимости: имена уровня модуля видны во всех
# функциях; для записи в них внутри функции нужно объявление `global`.
COUNTER = 0


def bump() -> None:
    global COUNTER
    COUNTER += 1
# --8<-- [end:req-8-2]


# --8<-- [start:variants-9]
# 9. Подпрограмму можно объявить в любом месте модуля, а также внутри
# другой подпрограммы — вложенная функция видит переменные внешней.
def outer(x: int) -> int:
    def inner(y: int) -> int:
        return x + y

    return inner(1)
# --8<-- [end:variants-9]


# --8<-- [start:variants-7]
# 7. Перегрузки по сигнатуре нет: второе `def` с тем же именем заменяет
# первое. Разное поведение для разных типов достигается проверкой типа
# внутри функции или functools.singledispatch.
def describe(value: object) -> str:
    if isinstance(value, int):
        return f"целое {value}"
    return f"объект {value!r}"
# --8<-- [end:variants-7]


# --8<-- [start:variants-8]
# 8. Передача параметров — по ссылке на объект (call by sharing):
# внутрь передаётся ссылка, изменение изменяемого объекта видно вызывающему,
# а перепривязка имени параметра — нет.
def append_and_rebind(items: list[int], n: int) -> None:
    items.append(n)   # видно снаружи: тот же объект list
    n = n + 1         # не видно снаружи: имя перепривязано локально
    items = []        # не видно снаружи: локальное имя указывает на новый list
# --8<-- [end:variants-8]


# --8<-- [start:req-4]
# 4. Ввод-вывод — встроенные функции input() и print(), импорт не нужен.
def read_int(prompt: str) -> int:
    return int(input(prompt))
# --8<-- [end:req-4]


def main() -> None:
    # --8<-- [start:variants-1]
    # 1. Объявление переменных неявное: переменная появляется при первом
    # присваивании. Аннотация типа (с 3.6) необязательна и не проверяется
    # интерпретатором.
    total = 0
    name: str = "ЯПИС"
    # --8<-- [end:variants-1]

    # --8<-- [start:variants-2]
    # 2. Преобразование типов — явное вызовом типа-конструктора; неявное
    # числовое преобразование есть в смешанной арифметике.
    # Проверка истинности использует отдельный протокол __bool__/__len__.
    count = int("42")
    ratio = count / 5.0        # смешанная арифметика int и float
    if count:                  # проверка истинности, count остаётся int
        total += count
    # --8<-- [end:variants-2]

    # --8<-- [start:variants-3]
    # 3. Присваивание одиночное и множественное (распаковка кортежа).
    a = 1
    a, b = b_first(), a        # правая часть вычисляется целиком до присваивания
    a = a + 1                 # перепривязка того же имени, не новое shadowing
    a = a - 1
    # --8<-- [end:variants-3]

    # --8<-- [start:variants-4]
    # 4. Области создают функции, классы, модули и comprehension в Python 3.
    # if/for/while новой области не создают; имя доступно после присваивания.
    if total > 0:
        inside_block = "видна ниже"
    print(inside_block)
    # --8<-- [end:variants-4]

    # --8<-- [start:variants-5]
    # 5. Маркер блока неявный: двоеточие и отступ, закрывающего символа нет.
    for i in range(3):
        total += i
        if i == 2:
            print("конец блока по отступу")
    # --8<-- [end:variants-5]

    # --8<-- [start:variants-6]
    # 6. Двухвариантный if / elif / else и, с версии 3.10, многовариантный
    # match — сопоставление с образцом, а не switch по константам.
    if total > 100:
        print("много")
    elif total > 10:
        print("средне")
    else:
        print("мало")

    match (a, b):
        case (1, _):
            print("первый элемент — единица")
        case (_, 1):
            print("второй элемент — единица")
        case _:
            print("что-то другое")
    # --8<-- [end:variants-6]

    # --8<-- [start:req-7-3]
    # 7.3. Цикл for — только по коллекции (foreach); счётный цикл — это
    # итерация по range().
    for ch in name:
        print(ch, end=" ")
    for i in range(0, 10, 2):
        print(i, end=" ")
    print()
    # --8<-- [end:req-7-3]

    # --8<-- [start:req-7-2-until]
    # 7.2. Отдельного until нет. Предусловие допускает ноль итераций;
    # постусловие repeat-until проверяется после тела, выполненного хотя бы раз.
    n = 0
    while not n >= 3:
        n += 1
    n = 0
    while True:
        n += 1
        if n >= 3:
            break
    # --8<-- [end:req-7-2-until]

    # --8<-- [start:req-7-2-do-while]
    # Цикла do-while нет; тело выполняется хотя бы раз через while True + break.
    while True:
        n -= 1
        if n <= 0:
            break
    # --8<-- [end:req-7-2-do-while]

    print(describe(count), describe(ratio), outer(41))
    items: list[int] = []
    append_and_rebind(items, 5)
    print(items)               # [5]
    bump()
    print(COUNTER)             # 1

    if len(sys.argv) > 1:
        print(read_int("число: "))


def b_first() -> int:
    return 2


if __name__ == "__main__":
    main()
